pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [check for columnHasSingleUniqueConstraint needs to handle hash-sharded primary key]

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkload [failed](https://tanzanite.cluster.engflow.com/invocations/default/466b90c3-1012-477f-a581-3314fd82cd47?testReportRun=24&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvd29ya2xvYWQvc2NoZW1hY2hhbmdlOnNjaGVtYWNoYW5nZV90ZXN0) on release-24.3 @ [de3dd55cfaa26de577b56f97b84c0e455815a074](https://github.com/cockroachdb/cockroach/commits/de3dd55cfaa26de577b56f97b84c0e455815a074):


```
    "query": "SELECT $1:::REGCLASS = $2:::REGCLASS",
    "queryArgs": [
     "public.table_w0_44",
     "public.table_w0_44"
    ],
    "result": true
   }
  ],
  "previousStatements": [
   "ALTER TABLE public.table_w0_44 ADD CONSTRAINT table_w0_44_col44_w0_45_table_w0_44_col44_w0_45_fk FOREIGN KEY (col44_w0_45) REFERENCES public.table_w0_44 (col44_w0_45) ON DELETE CASCADE ON UPDATE CASCADE"
  ],
  "usesDeclarativeSchemaChanger": true
 }
}
    schema_change_external_test.go:131: 
        	Error Trace:	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:131
        	Error:      	Received unexpected error:
        	            	***FAIL; Failed to receive an execution error when errors were expected
        	            	(1) forced error mark
        	            	  | "fatal error when running txn"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*opStmt).executeStmt
        	            	  | 	pkg/workload/schemachange/operation_generator.go:3223
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).runInTxn
        	            	  | 	pkg/workload/schemachange/schemachange.go:495
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	pkg/workload/schemachange/schemachange.go:592
        	            	  | pkg/ccl/testccl/workload/schemachange/schemachange_test_test.TestWorkload.TestWorkload.func4.func6
        	            	  | 	pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:120
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	external/org_golang_x_sync/errgroup/errgroup.go:78
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1695
        	            	Wraps: (4) ***FAIL; Failed to receive an execution error when errors were expected
        	            	Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.leafError
        	Test:       	TestWorkload
Schema Workload Stats
Total Schema Statements Executed = 50
Total Schema Statements Succeeded = 50
Total Schema Statement Expected Failures = 0
Total Transactions Committed = 35
Total Transactions Rolled Back = 21
Total Transactions Executed = 56
    schema_change_external_test.go:96: backup, tracing data, and system table dumps in /var/lib/engflow/worker/work/1/exec/bazel-out/k8-fastbuild/testlogs/pkg/ccl/testccl/workload/schemachange/schemachange_test/run_24_of_25/test.outputs/logTestWorkload2152290996
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestWorkload2152290996
--- FAIL: TestWorkload (11.65s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=24</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #137979 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [descriptor not found during DROP VIEW commit] [C-test-failure O-robot P-3 T-sql-foundations X-duplicate branch-release-23.2.19-rc]
- #137972 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [requested column not in index] [C-test-failure O-robot T-sql-foundations branch-release-24.3.3-rc release-blocker]
- #136966 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [duplicate key value violates unique constraint] [C-test-failure O-robot P-2 T-sql-foundations branch-master]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkload.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-45973

--- a/pkg/workload/schemachange/error_screening.go
+++ b/pkg/workload/schemachange/error_screening.go
@@ -9,6 +9,7 @@ import (
 	"context"
 	"fmt"
 	"regexp"
+	"slices"
 	"strings"
 
 	"github.com/cockroachdb/cockroach/pkg/sql/parser"
@@ -898,7 +899,8 @@ func (og *operationGenerator) constraintIsPrimary(
 	`, tableName.String()), constraintName)
 }
 
-// Checks if a column has a single unique constraint.
+// Checks if a column has a unique constraint or is in a primary key definition
+// that will guarantee uniqueness of the column.
 func (og *operationGenerator) columnHasSingleUniqueConstraint(
 	ctx context.Context, tx pgx.Tx, tableName *tree.TableName, columnName tree.Name,
 ) (bool, error) {
@@ -906,27 +908,60 @@ func (og *operationGenerator) columnHasSingleUniqueConstraint(
 	if columnName == "rowid" {
 		return true, nil
 	}
-	return og.scanBool(ctx, tx, `
-	SELECT EXISTS(
-	        SELECT column_name
-	          FROM (
-	                SELECT table_schema, table_name, column_name, ordinal_position,
-	                       concat(table_schema,'.',table_name)::REGCLASS::INT8 AS tableid
-	                  FROM information_schema.columns
-	               ) AS cols
-	          JOIN (
-	                SELECT contype, conkey, conrelid
-	                  FROM pg_catalog.pg_constraint
-	               ) AS cons ON cons.conrelid = cols.tableid
-	         WHERE table_schema = $1
-	           AND table_name = $2
-	           AND column_name = $3
-	           AND (contype = 'u' OR contype = 'p')
-	           AND array_length(conkey, 1) >= 1 -- unique index has to have this a prefix
-					   AND conkey[1] = ordinal_position
-	       )
-	`, tableName.Schema(), tableName.Object(), columnName)
+
+	rows, err := tx.Query(ctx, `
+	SELECT ordinal_position, conkey, shard_column_positions
+		FROM (
+					SELECT table_schema, table_name, column_name, ordinal_position, (SELECT array_agg(ordinal_position)
+									FROM information_schema.columns AS c
+									WHERE c.table_schema = columns.table_schema
+									AND c.table_name = columns.table_name
+									AND c.column_name ILIKE '%\_shard\_%')
+	        AS shard_column_positions, concat(table_schema,'.',table_name)::REGCLASS::INT8 AS tableid
+					FROM information_schema.columns
+				 ) AS cols
+		JOIN (
+					SELECT contype, conkey, conrelid
+						FROM pg_catalog.pg_constraint
+				 ) AS cons ON cons.conrelid = cols.tableid
+	 WHERE table_schema = $1
+		 AND table_name = $2
+		 AND column_name = $3
+		 AND (contype = 'u' OR contype = 'p')
+	`, tableName.Schema(), tableName.Object(), columnName,
+	)
+	if err != nil {
+		return false, err
+	}
+
+	isColUnique, err := pgx.CollectRows(rows, func(row pgx.CollectableRow) (bool, error) {
+		var ordinalPosition int
+		var conkey []int
+		var shardColumnPositions []int
+		if err := row.Scan(&ordinalPosition, &conkey, &shardColumnPositions); err != nil {
+			return false, err
+		}
+
+		// Find the first non-shard column in conkey, and check if it's equal
+		// the column we are inspecting.
+		for i, conkeyPos := range conkey {
+			if slices.Contains(shardColumnPositions, conkeyPos) {
+				continue
+			}
+			if conkeyPos == ordinalPosition {
+				// If this is the last column in the constraint, then it means the
+				// column is unique across the whole table.
+				return i == len(conkey)-1, nil
+			}
+		}
+		return false, nil
+	})
+	if err != nil {
+		return false, err
+	}
+	return slices.Contains(isColUnique, true), nil
 }
+
 func (og *operationGenerator) constraintIsUnique(
 	ctx context.Context, tx pgx.Tx, tableName *tree.TableName, constraintName string,
 ) (bool, error) {

